pkg/discovery/monitoring/kubelet/kubelet_monitor.go

```diff
@@ -0,0 +1,174 @@
+package kubelet
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+
+	api "k8s.io/client-go/pkg/api/v1"
+	stats "k8s.io/kubernetes/pkg/kubelet/apis/stats/v1alpha1"
+
+	"github.com/turbonomic/kubeturbo/pkg/discovery/metrics"
+	"github.com/turbonomic/kubeturbo/pkg/discovery/monitoring/types"
+	"github.com/turbonomic/kubeturbo/pkg/discovery/task"
+	"github.com/turbonomic/kubeturbo/pkg/discovery/util"
+
+	cadvisorapi "github.com/google/cadvisor/info/v1"
+
+	"github.com/golang/glog"
+)
+
+// KubeletMonitor is a resource monitoring worker.
+type KubeletMonitor struct {
+	nodeList []*api.Node
+
+	kubeletClient *kubeletClient
+
+	metricSink *metrics.EntityMetricSink
+
+	wg sync.WaitGroup
+}
+
+func NewKubeletMonitor(config *KubeletMonitorConfig) (*KubeletMonitor, error) {
+	kubeletClient, err := NewKubeletClient(config.KubeletClientConfig)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to create Kubelet client based on given config: %s", err)
+	}
+
+	return &KubeletMonitor{
+		kubeletClient: kubeletClient,
+		metricSink:    metrics.NewEntityMetricSink(),
+	}, nil
+}
+
+func (m *KubeletMonitor) GetMonitoringSource() types.MonitoringSource {
+	return types.KubeletSource
+}
+
+func (m *KubeletMonitor) ReceiveTask(task *task.Task) {
+	m.nodeList = task.NodeList()
+}
+
+func (m *KubeletMonitor) Do() *metrics.EntityMetricSink {
+	err := m.RetrieveResourceStat()
+	if err != nil {
+		glog.Errorf("Failed to execute task: %s", err)
+	}
+
+	return m.metricSink
+}
+
+// Start to retrieve resource stats for the received list of nodes.
+func (m *KubeletMonitor) RetrieveResourceStat() error {
+	if m.nodeList == nil || len(m.nodeList) == 0 {
+		return errors.New("Invalid nodeList or empty nodeList. Finish Immediately...")
+	}
+	m.wg.Add(len(m.nodeList))
+
+	for _, node := range m.nodeList {
+		// TODO, do we want to add timeout?
+		n := node
+		go m.scrapeKubelet(n)
```

Comment:
Performance concern:
should have a limit of the maximum number of concurrent goroutines.